crates/core/tedge_mapper/src/c8y/mapper.rs

```diff
@@ -108,3 +108,105 @@ fn create_directories() -> Result<(), anyhow::Error> {
+    const DEVICE_ID: &str = "test-device";
+    const DEVICE_NAME: &str = "test-user";
+    const DEVICE_TYPE: &str = "test-thin-edge.io";
+    const MQTT_HOST: &str = "127.0.0.1";
+
+    #[test_case(
+        "tedge/measurements",
+        "tedge/errors",
+        "{",
+        "Invalid JSON: EOF while parsing an object at line 1 column 1: `{`"
+    )] // fail case
+    #[test_case(
+        "tedge/measurements",
+        "tedge/measurements",
+        r#"{"temperature": 12, "time": "2021-06-15T17:01:15.806181503+02:00"}"#,
+        r#"{"temperature": 12, "time": "2021-06-15T17:01:15.806181503+02:00"}"#
+    )] // successs case
+    #[tokio::test]
+    async fn test_tedge_mapper_with_mqtt_pub(
+        pub_topic: &str,
+        sub_topic: &str,
+        payload: &str,
+        expected_msg: &str,
+    ) -> Result<(), anyhow::Error> {
+        // Mock endpoint to return C8Y internal id
+        let _get_internal_id_mock = mock("GET", "/identity/externalIds/c8y_Serial/test-device")
+            .with_status(200)
+            .with_body(
+                json!({ "externalId": DEVICE_ID, "managedObject": { "id": "123" } }).to_string(),
+            )
+            .create();
+
+        let mut jwt_token_retriver = Box::new(MockC8yJwtTokenRetriever::new());
+        jwt_token_retriver
+            .expect_get_jwt_token()
+            .returning(|| Ok(SmartRestJwtResponse::default()));
+
+        let http_client = reqwest::ClientBuilder::new().build().unwrap();
+        let proxy = JwtAuthHttpProxy::new(
+            jwt_token_retriver,
+            http_client,
+            mockito::server_url().as_str(),
+            DEVICE_ID,
+        );
+
+        // mapper config
+        let size_threshold = SizeThreshold(MQTT_MESSAGE_SIZE_THRESHOLD);
+        let operations = Operations::default();
+
+        let converter = Box::new(CumulocityConverter::new(
+            size_threshold,
+            DEVICE_NAME.into(),
+            DEVICE_TYPE.into(),
+            operations,
+            proxy,
+        ));
+
+        let broker = test_mqtt_broker();
+
+        let mut mapper = create_mapper(
+            CUMULOCITY_MAPPER_NAME_TEST,
+            MQTT_HOST.into(),
+            broker.port,
+            converter,
+        )
+        .await?;
+
+        // subscribe to `sub_topic`
+        let mut messages = broker.messages_published_on(sub_topic).await;
+
+        // run tedge_mapper in background
+        tokio::spawn(async move {
+            mapper
+                .run()
+                .instrument(info_span!(CUMULOCITY_MAPPER_NAME_TEST))
+                .await
+                .unwrap();
+        });
+
+        // publish `payload` to `pub_topic`
+        let () = broker.publish(pub_topic, payload).await?;
+
+        // check the `messages` returned contain `expected_msg`
+        assert_received_all_expected(&mut messages, TEST_TIMEOUT_SECS, &[expected_msg]).await;
+        Ok(())
```

Comment:
I hope that everything that could fail here will panic

Reply:
Yes, this is part of the Rust test crate